golang/x/relay/keeper/querier.go

```diff
@@ -58,24 +89,243 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
+	// check that the path is this many items long, error if not
+	if len(path) > 2 {
+		return []byte{}, types.ErrTooManyArguments(types.DefaultCodespace)
+	} else if len(path) < 2 {
+		return []byte{}, types.ErrNotEnoughArguments(types.DefaultCodespace)
+	}
+
+	hexDigest := path[0]
+	digestBytes, decodeErr := hex.DecodeString(hexDigest)
+	if decodeErr != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	digestLE, newDigestErr := btcspv.NewHash256Digest(digestBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	offset, convErr := strconv.Atoi(path[1])
+	if convErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, convErr)
+	}
+	newOffset := uint32(offset)
+
+	// This calls the keeper with the parsed arguments, and gets an answer
+	result, err := keeper.FindAncestor(ctx, digestLE, newOffset)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	// Now we format the answer as a response
+	response := types.QueryResFindAncestor{
+		DigestLE: digestLE,
+		Offset:   newOffset,
+		Res:      result,
+	}
+
+	// And we serialize that response as JSON
+	res, marshalErr := codec.MarshalJSONIndent(keeper.cdc, response)
+	if marshalErr != nil {
+		return []byte{}, types.ErrMarshalJSON(types.DefaultCodespace)
+	}
+	return res, nil
+}
+
+func queryHeaviestFromAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keeper Keeper) (res []byte, err sdk.Error) {
+	// check that the path is this many items long, error if not
+	if len(path) > 4 {
+		return []byte{}, types.ErrTooManyArguments(types.DefaultCodespace)
+	} else if len(path) < 4 {
+		return []byte{}, types.ErrNotEnoughArguments(types.DefaultCodespace)
+	}
+
+	ancestorHex := path[0]
+	ancestorBytes, decodeErr := hex.DecodeString(ancestorHex)
+	if decodeErr != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	ancestor, newDigestErr := btcspv.NewHash256Digest(ancestorBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	// parse the second path item as hex.
+	currentBest := path[1]
+	currentBestBytes, decodeErr := hex.DecodeString(currentBest)
+	if err != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	currentBestDigest, newDigestErr := btcspv.NewHash256Digest(currentBestBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	newBest := path[2]
+	newBestBytes, decodeErr := hex.DecodeString(newBest)
+	if err != nil {
+		return []byte{}, types.ErrBadHex(types.DefaultCodespace)
+	}
+	newBestDigest, newDigestErr := btcspv.NewHash256Digest(newBestBytes)
+	if newDigestErr != nil {
+		return []byte{}, types.FromBTCSPVError(types.DefaultCodespace, newDigestErr)
+	}
+
+	limit, convErr := strconv.Atoi(path[3])
```

Comment:
use `ParseUint` instead

also, if path is only 3 long, use a default number, as above